### PR TITLE
Feature scaling with law

### PR DESCRIPTION
**Pull request by lguzzi**
This PR allows to run the feature scaling task on condor using [law](https://github.com/riga/law)

### CODE REVIEW

File: Analysis/law/setup.sh [hunk 1]
```diff
@@ -1,17 +1,22 @@
 #!/usr/bin/env bash
 
 action() {
-    local this_file="$( [ ! -z "$ZSH_VERSION" ] && echo "${(%):-%x}" || echo "${BASH_SOURCE[0]}" )"
-    local this_dir="$( cd "$( dirname "$this_file" )" && pwd )"
+    eosfusebind -g
```
Comment (kandrosov) [inline Analysis/law/setup.sh:4]: Do we need to run `eosfusebind` also here? If yes, do the same check as in `bootstrap.sh`.

File: tau-ml-env.yaml [hunk 2]
```diff
@@ -328,6 +328,7 @@ dependencies:
   - zstandard=0.15.2=py38h497a2fe_0
   - zstd=1.4.9=ha95c52a_0
   - pip:
+    - law
```
Comment (kandrosov) [inline tau-ml-env.yaml:331]: please, specify the exact version as it is done for all other packages.

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Training/python/scaling_utils.py [hunk 3]
```diff
@@ -99,14 +99,14 @@ def init_dictionaries(features_dict, cone_selection_dict, n_files):
             assert len(var_dict) == 1
             (var, (_, _, scaling_type, *lim_params)), = var_dict.items()
             if scaling_type=='no_scaling' or scaling_type=='categorical':
-                scaling_params[var_type][var]['global'] = {"mean": 0, "std": 1, "lim_min": "-inf", "lim_max": "inf"}
+                scaling_params[var_type][var]['global'] = {"mean": 0, "std": 1, "lim_min": "-inf", "lim_max": "inf", "sqmean": None, "num": None}
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Training/python/scaling_utils.py:102]: I removed it by mistake when resoliving the merge conflicts (https://github.com/cms-tau-pog/TauMLTools/pull/82/commits/c80d33df88caee0fc9764a3deb11acd08b4fc2cd#diff-9773ab5333743c56a57cd9df63c241f34bc69ba57c4d7c386e1c0df38a730108L347) , thanks for spotting it.
Fixed in https://github.com/lguzzi/TauMLTools/commit/6faf68678a5086d1491a13b659e364a0dbf6ab92

File: Training/python/scaling_utils.py [hunk 4]
```diff
@@ -355,6 +362,11 @@ def fill_aggregators(tree, var, var_type, file_i, file_name_i, cone_type, cone_d
                 scaling_params[var_type][var][cone_type]['std'] = None
             else:
                 scaling_params[var_type][var][cone_type]['std'] = float(format(std_, '.4g'))
+            if sqmean_ == None:
+                print(f"Low statistics in {var} for sqmean computation")
+                scaling_params[var_type][var]['global']['sqmean'] = None
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Training/python/scaling_utils.py]: The bug was introduced by the conflict merge, thanks for spotting it
Fixed in https://github.com/lguzzi/TauMLTools/commit/91268d21fe49f29b8e5a76843e9bf73db24f5d3a

File: Training/python/merge_feature_scaling_jobs.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        assert all('sqmean' in j[vt][var][ct].keys() for j in jobs), "'sqmean' key not found for some of the jobs at "+thisstep
+        assert all('std'    in j[vt][var][ct].keys() for j in jobs), "'std' key not found for some of the jobs at "+thisstep
+        assert all('num'    in j[vt][var][ct].keys() for j in jobs), "'num' key not found for some of the jobs at "+thisstep
+
+        means   = [j[vt][var][ct]['mean']   for j in jobs]
+        sqmeans = [j[vt][var][ct]['sqmean'] for j in jobs]
+        stds    = [j[vt][var][ct]['std']    for j in jobs]
+        nums    = [j[vt][var][ct]['num']    for j in jobs]
+
+        assert all(check(v) for v in means  ), "'mean' value found 'nan' or 'inf' for some of the jobs at "+thisstep
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Training/python/merge_feature_scaling_jobs.py:61]: done
https://github.com/lguzzi/TauMLTools/commit/38b8e313b962690a58cf5363ebc1398a62377ff1

File: tau-ml-env.yaml (hunk 2 above)
Comment (lguzzi) [reply tau-ml-env.yaml]: done 
https://github.com/lguzzi/TauMLTools/commit/7646a23680c54bc82cec3ccb37f5aa086e7c01c7

**Review by yaourtpourtoi — CHANGES_REQUESTED** [review]

File: Training/python/scaling_utils.py [hunk 6]
```diff
@@ -344,7 +352,10 @@ def fill_aggregators(tree, var, var_type, file_i, file_name_i, cone_type, cone_d
         counts[var_type][var][cone_type][file_i] += ak.count(var_array[cone_mask])
         if fill_scaling_params:
             mean_ = compute_mean(sums[var_type][var][cone_type], counts[var_type][var][cone_type], aggregate=True)
+            sqmean_ =  compute_mean(sums2[var_type][var][cone_type], counts[var_type][var][cone_type], aggregate=True)
             std_ = compute_std(sums[var_type][var][cone_type], sums2[var_type][var][cone_type], counts[var_type][var][cone_type], aggregate=True)
+
+            scaling_params[var_type][var][cone_type]['num'] = int(counts[var_type][var].sum())
```
[the remark this answers is not in the document]
Comment (lguzzi) [reply Training/python/scaling_utils.py]: thanks, merged

**Review by kandrosov — APPROVED** [review]

**Pull request merged**